### PR TITLE
Makefile.include: creation of CACHEDIR.TAG as a dependency of pkg-prepare

### PR DESCRIPTION
**Pull request by bmewen**
### Contribution description

The cached dir, only build for now, are not created by default so I added a small default target that will create the build dir and the CACHEDIR whenever a make command is called.

### Testing procedure

Both commands should have the same behavior now. Previously pkg-prepare was not creating `build/CACHEDIR.TAG` and was creating the `build/` directory through `pkg/pkg.mk`.

```bash
make pkg-prepare
```

```bash
make
```

### Issues/PRs references

N/A

### CODE REVIEW

**Comment by bmewen** [comment]
@Enoch247

File: Makefile.include
Comment (kfessel) [inline Makefile.include:248]: such comfort targets are done by:

```
CREATECACHEDIR: $(BUILD_DIR)/CACHEDIR.TAG
```
just depend on what you want to have build

not sure if this case makes sense

Comment (Enoch247) [reply Makefile.include:248]: I concur with the above.

File: Makefile.include
Comment (Enoch247) [inline Makefile.include:253]: This is a novel/interesting use of the include directive. You may consider adding a comment explaining the intent.

**Review by Enoch247 — COMMENTED** [review]
You might consider taking a look at [order only prerequisites](
https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html) to accomplish your goal. 

You could leave the existing rule to create the tag file as-is. Then just add it as an order-only-prereq to any rules that should create the build dir.

File: Makefile.include
Comment (bmewen) [reply Makefile.include:253]: I saw what you said earlier about the `order only prerequisites` but I don't think it's the best way to handle this as it means that every new target that will be added in the future will need to ensure it includes this specific target as a prequisite.

On the opposite side what I've done using `-include CREATECACHEDIR` will call the target `CREATECACHEDIR` regardless of the original target called by the user which is the use case we've identified so far. Even if it's probably not the cleanest way to do that I think that the maintainability is a really strong argument to its usage.

File: Makefile.include
Comment (bmewen) [reply Makefile.include]: I edited it to keep the `$(BUILD_DIR)/CACHEDIR.TAG` target. 

And no, I'm not replacing the target `all`. The includes means that regardless of the target called originally, the `CREATECACHEDIRS` will also be called by default.

The problem with using `all` is that you have to specify `CREATECACHEDIRS` as a dependency for every single target that will use it. (which is very uncomfortable for me tbh)

Comment (kfessel) [reply Makefile.include]: The default is not about the include but about being the first non dot rule in a makefile which is the default if you call just `$ make ` previously it was `all`, which also include with you rule above it it is create a directory.

the include is just a strange way to enforce something happening (a hopefully allways failing try of an include)

https://www.gnu.org/software/make/manual/make.html#How-Make-Works

> By default, make starts with the first target (not targets whose names start with ‘.’ unless they also contain one or more ‘/’). This is called the default goal. 

https://www.gnu.org/software/make/manual/make.html#Standard-Targets
> All GNU programs should have the following targets in their Makefiles: 
> ‘all’ ... Compile the entire program. This should be the default target.

There probably is a reason you are doing these strange things so maybe you can enlighten me -- Why is it uncomfotable to have the directory created when it is needed?

File: Makefile.include
Comment (kfessel) [inline Makefile.include:248]: this should probably not be the first rule / default target

File: Makefile.include
Comment (kfessel) [reply Makefile.include]: its probably a good idea to create the directory when you need it -> depend on its or the tag file existence

Comment (kfessel) [reply Makefile.include]: Ahh i see it is the `mkdir -p` in the pkg.mk which create the directory without the tag 

probably the right solution would be to for `$(PKG_DOWNLOADED):` to depend on the CACHEDIR.TAG file  after checking if the state-files share a part of the path with the CACHEDIR.TAG

File: Makefile.include
Comment (bmewen) [reply Makefile.include]: As I said, it could be added in the targets that needs a cache directory and work as well but it means that whenever a target would need this dependency it will need to include it in its prerequisites. I don't think that creating empty directories in advance (that will most likely be filled soon) is costing that much. 

Also the case I mentionned, `pkg-build` was using `$(BUILD_DIR)/CACHEDIR.TAG` but not `pkg-prepare` when it was the target that actually fills the directory. With this specific case in mind I thought it was better to call `CREATECACHEDIRS` once by default rather than calling it twice (once in each target).

File: Makefile.include
Comment (bmewen) [reply Makefile.include]: I think that we also previously discussed the fact that there might also be some other cache directories. Like other directories with a CACHEDIR.TAG that are not build, which motivated my choice of creating a new target for this specific job. It means that whenever a new directory with a CACHEDIR.TAG will be needed it will only need to be added as a prerequisite in the target I created.

**Comment by bmewen** [comment]
I am still waiting for a review whenever someone is available. Thank you

**Comment by kfessel** [comment]
> I am still waiting for a review whenever someone is available. Thank you

From my POV this PR is a strange workaround abusing `include` for a minor problem (not many people call pkg-prepare directly)

- the include will search for the file at more than one location and include it if found https://www.gnu.org/software/make/manual/make.html#Include `First, any directories you have specified with the ‘-I’ or ‘--include-dir’ options are searched (see [Summary of Options](https://www.gnu.org/software/make/manual/make.html#Options-Summary)). Then the following directories (if they exist) are searched, in this order: prefix/include (normally /usr/local/include [1](https://www.gnu.org/software/make/manual/make.html#FOOT1)) /usr/gnu/include, /usr/local/include, /usr/include.`

- you identified one rule where the file should be generated but is not - there is a more straight forward fix for that ( just add a requirement to that rule  

- searching for `mkdir` -p or `${BUILD_DIR}` did not reveal an unmanageable number of instances that might need fixing in future

So in this form i don't think it is good to go (no positive from me)
but i also write this review as a comment -> easy to ignore (no negative)

**Review by Enoch247 — CHANGES_REQUESTED** [review]
The way you have used the `include` here is really clever, but not the right tool for the job in my opinion. The problem with doing it this way is that it is *always* run for every possible invocation of make even invocations that should not create anything such as any of the `info-*` recipes. This could be a bit of a surprise in environments when write access is not expected. Also, this causes the build dir to be create always, even if not needed (for example `make BOARD=native examples/blinky`).

**Review by kfessel — APPROVED** [review]
Thank you for your contribution

**Comment by Enoch247** [comment]
Thanks, I will test and review this weekend.

**Review by Enoch247 — CHANGES_REQUESTED** [review]
I would prefer to see the `prepare` target inside of `pkg.mk` depend on `$(BUILD_DIR)/CACHEDIR.TAG`, or the `prepare` target depending on '$(PKGDIRBASE)' and '$(PKGDIRBASE)' on `$(BUILD_DIR)/CACHEDIR.TAG`, but I won't insist on it.

If you decide to keep the current approach, please accept the suggested change for line 807.

File: Makefile.include
Comment (bmewen) [reply Makefile.include:807]: I can definitely move `$(BUILD_DIR)/CACHEDIR.TAG` in the `prepare` target but `$(PKGDIRBASE)` is not a target, hence I can't really add a dependency on it

**Comment by bmewen** [comment]
I genuinely don't think that this should be normal. Bossa is not a package so why should it rely on `pkg/pkg.mk`, furthermore why is built in the `build/pkg` directory. That's why all the checks fail and I believe this behavior is the same for all the `dist/tools` builds that failed.
image

**Comment by bmewen** [comment]
@Enoch247 this version is good for me if you have time to review it once more.

**Comment by Enoch247** [comment]
Thank you. I will take a look tomorrow.

File: pkg/pkg.mk
Comment (Enoch247) [inline pkg/pkg.mk:85]: I feel this recipe should be placed outside of this file, as it is at a broader scope than building packages. Perhaps as an include-able makefile in `./makefiles`. However, I will approve as-is.

**Review by Enoch247 — APPROVED** [review]

**Pull request merged**
